arrow-libs/fx/arrow-fx-resilience/src/commonMain/kotlin/arrow/fx/resilience/Schedule.kt

```diff
@@ -1,1091 +1,500 @@
- * }
- * ```
- * <!--- KNIT example-schedule-06.kt -->
- *
- * ## Repeating an effect until/while it produces a certain value
- *
- * We can make use of the policies doWhile or doUntil to repeat an effect while or until its produced result matches a given predicate.
- *
- * ```kotlin
- * import arrow.fx.resilience.*
- *
- * suspend fun main(): Unit {
- *   var counter = 0
- *   //sampleStart
- *   val res = Schedule.doWhile<Int>{ it <= 3 }.repeat {
- *     println("Run: ${counter++}"); counter
- *   }
- *   //sampleEnd
- *   println(res)
- * }
- * ```
- * <!--- KNIT example-schedule-07.kt -->
- *
- * ## Exponential backoff retries
- *
- * A common algorithm to retry effectful operations, as network requests, is the exponential backoff algorithm. There is a scheduling policy that implements this algorithm and can be used as:
- *
- * ```kotlin
- * import kotlin.time.Duration.Companion.milliseconds
- * import kotlin.time.ExperimentalTime
- * import arrow.fx.resilience.*
- *
- * @ExperimentalTime
- * val exponential = Schedule.exponential<Unit>(250.milliseconds)
- * ```
- * <!--- KNIT example-schedule-08.kt -->
+ * It's defined by a [step] function that takes an [Input] and returns a [Decision],
+ * the [Decision] determines if the `suspend fun` should be [Continue] to be retried or repeated,
+ * or if the [Schedule] is [Done] retrying or repeating.
  */
-public sealed class Schedule<Input, Output> {
-
-  public abstract suspend fun <C> repeatOrElseEither(
-    fa: suspend () -> Input,
-    orElse: suspend (Throwable, Output?) -> C
-  ): Either<C, Output>
-
-  /**
-   * Runs this effect once and, if it succeeded, decide using the provided policy if the effect should be repeated and if so, with how much delay.
-   * Returns the last output from the policy or raises an error if a repeat failed.
-   */
-  public suspend fun repeat(fa: suspend () -> Input): Output =
-    repeatOrElse(fa) { e, _ -> throw e }
-
-  /**
-   * Runs this effect once and, if it succeeded, decide using the provided policy if the effect should be repeated and if so, with how much delay.
-   * Also offers a function to handle errors if they are encountered during repetition.
-   */
-  public suspend fun repeatOrElse(fa: suspend () -> Input, orElse: suspend (Throwable, Output?) -> Output): Output =
-    repeatOrElseEither(fa, orElse).fold(::identity, ::identity)
-
-  public abstract suspend fun <C> repeatOrElseEitherAsFlow(
-    fa: suspend () -> Input,
-    orElse: suspend (Throwable, Output?) -> C
-  ): Flow<Either<C, Output>>
-
-  /**
-   * Runs this effect and emits the output, if it succeeded, decide using the provided policy if the effect should be repeated and emitted, if so, with how much delay.
-   * This will raise an error if a repeat failed.
-   */
-  public suspend fun repeatAsFlow(fa: suspend () -> Input): Flow<Output> =
-    repeatOrElseAsFlow(fa) { e, _ -> throw e }
+@JvmInline
+public value class Schedule<Input, Output>(public val step: Next<Input, Output>) {
```

Comment:
Maybe something like `DecisionInput` instead of simply `Input`. My reason is that a schedule needs some input to make its decision, and in things like `repeat` this input is simply the output of the function. I would still change the signature to:

```kotlin
suspend fun <A> Schedule<Throwable, *>.repeat(action: suspend () -> A): A
```

to make it more clear.

Reply:
🤕 isn't that the `retry` signature, and it's already defined as such in the PR.

Reply:
No, `retry` uses `Input` instead of `A` (using `Input` there is what I find confusing)

Reply:
Aah, okay. Now I understood, yes using `Input` for `retry` is indeed confusing. That was an incorrect refactor.